src/shared/varlink.c

```diff
@@ -1219,18 +1220,19 @@ int varlink_close(Varlink *v) {
 Varlink* varlink_close_unref(Varlink *v) {
         if (!v)
                 return NULL;
-
-        (void) varlink_close(v);
+                                   /* n_ref >= 1 */
+        varlink_ref(v);            /* n_ref >= 2 */
+        varlink_close(v);          /* n_ref >= 1 */
+        if (v->n_ref > 1)
+                v->n_ref--;        /* n_ref >= 1 */
```

Comment:
i don't get this? can you elaborate? now you sometimes drop a reference and sometimes you don't? that's looks very wrong to me?

Reply:
The sequence of events is:
when we enter `varlink_close_unref`, n_ref==1. We call varlink_close() which drops the last reference from a callback, so n_ref==0 and the object is deallocated. We then call varlink_unref() ourselves, and crash on access to freed memory. By holding the reference to the object ourselves, we know the object is not deallocated. Essentially, we're saying: we take a reference and will drop it, unless the varlink_close() call we call happens to drop the next-to-last reference, in which case we'll just gracefully deallocate the object before returning.

We could say that freeing the object from the callback is not allowed, but that would be rather unpleasant.

Reply:
Hmm, but how can n_ref==1 even be when entering varlink_close_unref() ?